### PR TITLE
Recognize Match Statement

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**
Fixes: #1022

### CODE REVIEW

**Comment by certik** [comment]
Here are some more test cases to try:
```
def case(x):
    pass

def match(x):
    pass
```

File: src/lpython/parser/tokenizer.re [hunk 1]
```diff
@@ -397,6 +397,10 @@ int Tokenizer::lex(Allocator &al, YYSTYPE &yylval, Location &loc, diag::Diagnost
             "yield"    { KW(YIELD) }
             "yield" whitespace "from" whitespace { KW(YIELD_FROM) }
 
+            // Soft Keywords
+            "match" / [^\n\x00]+ ":" newline { KW(MATCH) }
+            "case" / [^\n\x00]+ ":" newline { KW(CASE) }
```
Comment (certik) [inline src/lpython/parser/tokenizer.re:402]: I think to fix the test cases that I posted above, you need to check here that the last token was TK_NEWLINE, if so, it's a KW_MATCH, otherwise it's a TK_NAME.

**Comment by certik** [comment]
For `case` here is a good one:
```python
case = {"ok": 5}
a = {
case ["ok"]:
  6}
print(a)
```
I don't think you can even distinguish this at the token level.

**Comment by certik** [comment]
Let's try this:

* Must be immediately preceded by TK_NEWLINE and optional indent (for match) or required indent (for case)
* the line must end with : and \n
* parenlevel must be 0

That might do it.

**Comment by certik** [comment]
Hm, but what about this:
```
>>> def file_handler_v1(command):
...     match command.split():
...         case ['show']:
...             print('List all files and directories: ')
...             # code to list files
...         case ['remove',
   *files]:
...             print('Removing files: {}'.format(files))
...             # code to remove files
...         case _:
...             print('Command not recognized')
```
The second `case` does not end with `:`...

**Comment by Thirumalai-Shaktivel** [comment]
Exactly, I'm working on it.

File: tests/parser/match_stmt1.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        case [*_]:
+            return "seq"
+
+def file_handler_v1(command):
+    match command.split():
+        case ['show']:
+            print('List all files and directories: ')
+            # code to list files
+        case ['remove',
+              *files]:
```
Comment (Thirumalai-Shaktivel) [inline tests/parser/match_stmt1.py:33]: Fixed all the cases, but this is tricky. Here, the parentheses value is not recorded, simply put the `parenlevel = 0`.

Comment (certik) [reply tests/parser/match_stmt1.py:33]: Why is it not recorded?

**Comment by certik** [comment]
Another use case:
```python
match = """some string:
  surprise!"""
```

File: tests/parser/match_stmt1.py (hunk 2 above)
Comment (certik) [reply tests/parser/match_stmt1.py:33]: Ah I see, it's not recorded in the lookahead part. 

Yes, indeed, that's because regex cannot parse nested parentheses.

**Comment by certik** [comment]
I think I would try the following. When you see "match space character" (and you are not inside a string), then it must be a keyword match, isn't it?

If so, when you see this, call a function `lex_match()`. Inside `lex_match`, finishing tokenizing the rest of the line, until a TK_NEWLINE. Then check on the new line that it is indented and it starts with "case". If so, call lex_case() and repeat. If it does not start with "case", that's the end of match, so return.

We can iterate on the exact details of lex_match() and lex_case(), but the overall structure seems very robust.

**Comment by Thirumalai-Shaktivel** [comment]
I will try to implement this, thank you.

**Comment by Thirumalai-Shaktivel** [comment]
I'm confused here:
- the `lex_match()` and `lex_case()` are a mini-tokenizer which is used to recognize the match statement, right?
- if so, how would I return each recognized tokens to `tokens()` function?
- Also, should I re-define everything like `digits`, `integer`, `real`, `string`... ?
- can `lex()` function be called inside `lex_match()`?
- Everything works but only the `strings` and `parentheses` to be fixed.

**Comment by certik** [comment]
Let's do the following. Your current design is good enough (except strings and parentheses), so let's write test, polish this up and merge it.

In a subsequent PRs we'll try to fix the remaining case of strings and parentheses. Yes, it's tough to return the tokens from `lex_match()`. One idea is to only make `lex_match/lex_case()` "parse" (say using re2c) but without returning tokens, just to recognize robustly if it is indeed a "match" case. Then come back to the main tokenizer and return KW_MATCH and continue matching as usual, until a "case" is encountered. We can then use a "match state" and beginning of the line / indent to robustly ensure we have a "case" (I think). Or something like that.

**Comment by Thirumalai-Shaktivel** [comment]
I got it, thanks.

**Comment by Thirumalai-Shaktivel** [comment]
This PR is ready to go in.

**Review by certik — APPROVED** [review]
I think this is excellent, thanks for getting most of it done. The tests look thorough. There are some corner cases (commented out) that doesn't work, but this should be good enough to get started. 

Can you open up an issue for the cases that don't work yet?

**Comment by Thirumalai-Shaktivel** [comment]
Thank you for reviewing and merging this PR.
Yup, I will create an issue for it.
